File: lib/bullx/ai_agent/tools/web/exa.ex
defmodule BullX.AIAgent.Tools.Web.Exa do
  @moduledoc """
  Exa adapter for BullX AIAgent web search and extraction tools.

  The adapter translates Exa's response fields into BullX's normalized
  `title`/`url`/`snippet` or extracted-text result shape.
  """

  alias BullX.AIAgent.Tools.Web

  @search_url "https://api.exa.ai/search"
  @contents_url "https://api.exa.ai/contents"

  def available?(_kind), do: not is_nil(Web.api_key(:exa))

  def search(%{query: query, limit: limit}, runtime_seed) do
    body = %{
      "query" => query,
      "type" => "auto",
      "numResults" => limit,
      "contents" => %{"highlights" => true}
    }

    [url: @search_url, json: body, headers: auth_headers(), retry: false]
    |> Keyword.merge(Web.req_options(runtime_seed))
    |> Req.post()
    |> Web.http_result()
    |> case do
      {:ok, response} ->
        {:ok,
         %{"success" => true, "query" => query, "results" => search_results(response, limit)}}

      {:error, error} ->
        {:error, error}
    end
  end

  def extract(%{urls: urls}, runtime_seed) do
    body = %{"urls" => urls, "text" => true}

    [url: @contents_url, json: body, headers: auth_headers(), retry: false]
    |> Keyword.merge(Web.req_options(runtime_seed))
    |> Req.post()
    |> Web.http_result()
    |> case do
      {:ok, response} -> {:ok, %{"success" => true, "results" => extract_results(response)}}
      {:error, error} -> {:error, error}
    end
  end

  defp auth_headers, do: [{"x-api-key", Web.api_key(:exa)}]

  defp search_results(%{"results" => results}, limit) when is_list(results) do
    results
    |> Enum.take(limit)
    |> Enum.map(&search_result/1)
  end

  defp search_results(_response, _limit), do: []

  defp search_result(%{} = item) do
    %{
      "title" => string_value(item, "title"),
      "url" => string_value(item, "url"),
      "snippet" => exa_snippet(item)
    }
    |> maybe_put_position(item)
  end

  defp exa_snippet(%{"highlights" => highlights}) when is_list(highlights) do
    highlights
    |> Enum.filter(&is_binary/1)
    |> Enum.join("\n")
  end

  defp exa_snippet(%{"text" => text}) when is_binary(text), do: text
  defp exa_snippet(%{"snippet" => snippet}) when is_binary(snippet), do: snippet
  defp exa_snippet(_item), do: ""

  defp extract_results(%{"results" => results} = response) when is_list(results) do
    status_by_url = status_by_url(Map.get(response, "statuses"))

    Enum.map(results, fn item ->
      url = string_value(item, "url")

      %{
        "url" => url,
        "title" => string_value(item, "title"),
        "text" => string_value(item, "text")
      }
      |> maybe_put_error(Map.get(status_by_url, url))
    end)
  end

  defp extract_results(_response), do: []

  defp status_by_url(statuses) when is_list(statuses) do
    statuses
    |> Enum.filter(&is_map/1)
    |> Map.new(fn status -> {string_value(status, "url"), status} end)
  end

  defp status_by_url(_statuses), do: %{}

  defp maybe_put_error(result, %{"status" => status}) when status not in ["success", "ok", 200],
    do: Map.put(result, "error", to_string(status))

  defp maybe_put_error(result, _status), do: result

  defp maybe_put_position(result, %{"position" => position}) when is_integer(position),
    do: Map.put(result, "position", position)

  defp maybe_put_position(result, _item), do: result

  defp string_value(map, key) do
    case Map.get(map, key) || Map.get(map, String.to_existing_atom(key)) do
      value when is_binary(value) -> value
      value when is_integer(value) or is_float(value) -> to_string(value)
      _value -> ""
    end
  rescue
    ArgumentError -> Map.get(map, key) || ""
  end
end

File: app/src/ai-agent/web/providers/exa.ts
import { appConfigService } from '@/config/app-configure'
import { WebExaApiKey } from '../config'
import { normalizeUrl, requestJson } from '../http'
import {
  type WebExtractArgs,
  type WebExtractResult,
  type WebProvider,
  WebProviderError,
  type WebSearchArgs,
  type WebSearchResult
} from '../provider'

const SEARCH_URL = 'https://api.exa.ai/search'
const CONTENTS_URL = 'https://api.exa.ai/contents'

interface ExaSearchResponse {
  results?: Array<{ title?: string; url?: string; text?: string; highlights?: string[]; summary?: string }>
}

interface ExaContentsResponse {
  results?: Array<{ id?: string; url?: string; title?: string; text?: string }>
  statuses?: Array<{ id?: string; status?: string; error?: { tag?: string } }>
}

async function apiKey(): Promise<string | undefined> {
  return appConfigService.get(WebExaApiKey)
}

function requireKey(key: string | undefined): string {
  if (!key) throw new WebProviderError('exa api key not configured', { retryable: false, providerId: 'exa' })
  return key
}

/**
 * Exa adapter — the only built-in that serves both capabilities. Search hits
 * `/search`, extraction hits `/contents`. Auth is the `x-api-key` header, and a
 * missing key makes the provider report unavailable (no key => `available()`
 * false), so the registry skips it during fallback instead of failing a call.
 */
export const exaProvider: WebProvider = {
  id: 'exa',
  supports: ['search', 'extract'],
  async available() {
    return Boolean(await apiKey())
  },
  async unavailableReason() {
    return (await apiKey()) ? undefined : 'exa api key not configured'
  },
  async search(args: WebSearchArgs, signal?: AbortSignal): Promise<WebSearchResult[]> {
    const key = requireKey(await apiKey())
    const limit = args.limit ?? 5
    const data = await requestJson<ExaSearchResponse>('exa', SEARCH_URL, {
      method: 'POST',
      headers: { 'content-type': 'application/json', 'x-api-key': key },
      body: JSON.stringify({ query: args.query, type: 'auto', numResults: limit, contents: { highlights: true } }),
      signal
    })
    // Exa may populate any of several snippet fields per hit. Prefer the
    // query-relevant `highlights` (we asked for them), then a `summary`, then the
    // raw `text`, falling to '' so the normalized shape always has a snippet.
    // `slice(limit)` re-applies the cap defensively in case the API over-returns.
    return (data.results ?? []).slice(0, limit).map(result => ({
      title: result.title ?? '',
      url: result.url ?? '',
      snippet:
        (Array.isArray(result.highlights) ? result.highlights.join(' ') : undefined) ??
        result.summary ??
        result.text ??
        ''
    }))
  },
  async extract(args: WebExtractArgs, signal?: AbortSignal): Promise<WebExtractResult[]> {
    const key = requireKey(await apiKey())
    const data = await requestJson<ExaContentsResponse>('exa', CONTENTS_URL, {
      method: 'POST',
      headers: { 'content-type': 'application/json', 'x-api-key': key },
      body: JSON.stringify({ urls: args.urls, text: true }),
      signal
    })
    // Exa returns content and per-URL failure in two parallel arrays (`results`
    // and `statuses`), keyed by url/id and not guaranteed to match the request
    // order. Index both by a normalized URL so we can re-join them back onto the
    // caller's exact input list below. `normalizeUrl` absorbs trailing-slash /
    // whitespace drift between what we sent and what Exa echoes back.
    const byUrl = new Map((data.results ?? []).map(result => [normalizeUrl(result.url ?? result.id ?? ''), result]))
    const errorByKey = new Map(
      (data.statuses ?? [])
        .filter(status => status.status && status.status !== 'success')
        .map(status => [normalizeUrl(status.id ?? ''), status.error?.tag ?? 'extract failed'])
    )
    // Project onto the input URLs so the result is 1:1 with the request (same
    // length and order the caller expects). A URL with no content row becomes a
    // per-URL error — the matching status tag if Exa reported one, else a generic
    // "no content returned" — rather than being dropped.
    return args.urls.map(url => {
      const key = normalizeUrl(url)
      const result = byUrl.get(key)
      if (result) return { url, title: result.title ?? '', text: result.text ?? '' }
      return { url, title: '', text: '', error: errorByKey.get(key) ?? 'no content returned' }
    })
  }
}
